File: problems/树/103.二叉树的锯齿形层序遍历.py
#!/usr/bin/env python3
"""
103. 二叉树的锯齿形层序遍历
"""
from typing import List, Optional
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def zigzag_level_order(root: Optional[TreeNode]) -> List[List[int]]:
    if not root: return []
    result = []
    queue = deque([root])
    reverse = False
    while queue:
        level = []
        for _ in range(len(queue)):
            node = queue.popleft()
            level.append(node.val)
            if node.left:  queue.append(node.left)
            if node.right: queue.append(node.right)
        if reverse: level.reverse()
        result.append(level)
        reverse = not reverse
    return result

if __name__ == "__main__":
    root1 = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    print(zigzag_level_order(root1))  # [[3], [20, 9], [15, 7]]
    print(zigzag_level_order(TreeNode(1)))  # [[1]]
    print(zigzag_level_order(None))  # []

File: problems/树/103.二叉树的锯齿形层序遍历.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
    if (!root) return {};
    vector<vector<int>> result;
    queue<TreeNode*> q;
    q.push(root);
    bool rev = false;
    while (!q.empty()) {
        int sz = q.size();
        vector<int> level;
        while (sz--) {
            TreeNode* node = q.front(); q.pop();
            level.push_back(node->val);
            if (node->left)  q.push(node->left);
            if (node->right) q.push(node->right);
        }
        if (rev) reverse(level.begin(), level.end());
        result.push_back(level);
        rev = !rev;
    }
    return result;
}
int main() {
    TreeNode* root = new TreeNode(3);
    root->left  = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left  = new TreeNode(15);
    root->right->right = new TreeNode(7);
    auto r = zigzagLevelOrder(root);
    for (auto& lv : r) { for (int v : lv) cout << v << " "; cout << endl; }
    // [[3], [20, 9], [15, 7]]
    r = zigzagLevelOrder(new TreeNode(1));
    for (int v : r[0]) cout << v << " "; cout << endl;  // 1
    cout << zigzagLevelOrder(nullptr).size() << endl;  // 0
    return 0;
}
